src/middlewared/middlewared/client/client.py

```diff
@@ -48,126 +44,55 @@ def wait(self, timeout=None):
-
-                    socktype = socket.SOCK_STREAM
-                    proto = 0
-
-                """
-                This is the line replaced to use socket.fromfd
-                """
-                try:
-                    self.reserved_fd = self.get_reserved_portfd(blacklist=reserved_ports_blacklist)
-                    sock = socket.fromfd(self.reserved_fd, family, socktype, proto)
-                    sock.setsockopt(socket.IPPROTO_TCP, socket.TCP_NODELAY, 1)
-                    sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-                    if hasattr(socket, 'AF_INET6') and family == socket.AF_INET6 and self.host.startswith('::'):
-                        try:
-                            sock.setsockopt(socket.IPPROTO_IPV6, socket.IPV6_V6ONLY, 0)
-                        except (AttributeError, socket.error):
-                            pass
-                except Exception as e:
-                    if self.reserved_fd:
-                        try:
-                            os.close(self.reserved_fd)
-                        except OSError:
-                            pass
-                    raise e
-
-            WebSocket.__init__(self, sock, protocols=kwargs.get('protocols'),
-                               extensions=kwargs.get('extensions'),
-                               heartbeat_freq=kwargs.get('heartbeat_freq'))
-
-            self.stream.always_mask = True
-            self.stream.expect_masking = False
-            self.key = b64encode(os.urandom(16))
-            self._th = threading.Thread(target=self.run, name='WebSocketClient')
-            self._th.daemon = True
-
-    def get_reserved_portfd(self, blacklist=None):
+            self.privileged_ports_blacklist = list(set(self.privileged_ports_blacklist + [860]))
+
         """
-        Get a file descriptor with a reserved port (<=1024).
-        The port is arbitrary using the libc "rresvport" call and its tried
-        again if its within `blacklist` list.
+        This gets the IP address of the internal NIC on TrueNAS HA systems.
         """
-        if blacklist is None:
-            blacklist = []
-
-        libc = ctypes.cdll.LoadLibrary('libc.so.7')
-        port = ctypes.c_int(0)
-        pport = ctypes.pointer(port)
-        fd = libc.rresvport(pport)
-        retries = 5
-        while True:
-            if retries == 0:
-                break
-            if fd < 0:
-                time.sleep(0.1)
-                fd = libc.rresvport(pport)
-                retries -= 1
-                continue
-            if pport.contents.value in blacklist:
-                oldfd = fd
-                fd = libc.rresvport(pport)
-                os.close(oldfd)
-                retries -= 1
+        host = ''
+        try:
+            with socket.socket(socket.AF_INET, socket.SOCK_DGRAM) as s:
+                s.connect(('169.254.10.1', 80))
+                host = s.getsockname()[0]
+        except Exception:
+            pass
```

Comment:
Seems dubious to `pass` here.

Reply:
I pass there under the premise that we don't have to have the local IP address for which we `.bind()` to as it will be built based on the URI parameter that get's called in the failover plugin. What would you suggest we do in that scenario? Log it, maybe?

Reply:
Oh I actually missed that you have `host=''` above this.
```python
import socket
s = socket.socket()
s.bind(('', 123))
s.connect(('localhost', 22))
while True:
    pass
```
```
# bind.py &
# sockstat | grep 123   
root     python3.6  15498 3  tcp4   127.0.0.1:123         127.0.0.1:22
?        ?          ?     ?  tcp4   127.0.0.1:22          127.0.0.1:123
```
I just tested the above code which shows we can bind to a wildcard address and only set a specific port. The kernel will pick the correct source address when we connect based on the address we connect to. So we don't benefit from this extra socket work here.

Reply:
Even better!